flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java

```diff
@@ -94,26 +130,218 @@ public RegistrationResponse apply(final JobMasterGateway jobMasterGateway) {
+			// TODO slotManager.notifyTaskManagerFailure(resourceID);
+			taskManager.shutDown(leaderSessionID);
+		}
+	}
+
+	/**
+	 * close heartbeat triggers to resource if exist
+	 * @param resourceID which resource need to stop keep heartbeat with
+	 */
+	private void closeHeartbeatToResourceIfExist(ResourceID resourceID) {
+		if(heartbeatSchedulers.containsKey(resourceID)) {
+			ResourceManagerToTaskExecutorHeartbeatScheduler heartbeatManager = heartbeatSchedulers.get(resourceID);
+			heartbeatManager.close();
+			heartbeatSchedulers.remove(resourceID);
+		}
+	}
+
+	/**
+	 * send slotRequest to the taskManager which the given slot is on
+	 *
+	 * @param slotRequest slot request information
+	 * @param slotID      which slot is choosen
+	 */
+	void requestSlotToTaskManager(final SlotRequest slotRequest, final SlotID slotID) {
+		ResourceID resourceID = slotID.getResourceID();
+		TaskExecutorGateway taskManager = taskExecutorGateways.get(resourceID);
+		if (taskManager == null) {
+			// the given slot is on an unregister taskManager
+			log.warn("ignore slot {} because it is on an unregister taskManager", slotID);
+			// TODO slotManager.handleSlotRequestFailedAtTaskManager(slotRequest, slotID);
+		} else {
+			Future<SlotAllocationResponse> response = taskManager.requestSlotForJob(
+				slotRequest.getAllocationID(), slotRequest.getJobID(), slotID, leaderSessionID);
+			response.onSuccess(new OnSuccess<SlotAllocationResponse>() {
+				@Override
+				public void onSuccess(SlotAllocationResponse result) throws Throwable {
+					if (result instanceof SlotAllocationResponse.Decline) {
+						// TODO slotManager.handleSlotRequestFailedAtTaskManager(slotRequest, slotID);
+					}
+				}
+			}, getMainThreadExecutionContext());
+			response.onFailure(new OnFailure() {
+				@Override
+				public void onFailure(Throwable failure) {
+					log.error("fail to request slot on taskManager because of error", failure);
+					// TODO slotManager.handleSlotRequestFailedAtTaskManager(slotRequest, slotID);
+				}
+			}, getMainThreadExecutionContext());
+		}
+	}
+
+	/**
+	 * notify slotReport which is sent by taskManager to resourceManager
+	 *
+	 * @param slotReport the slot allocation report from taskManager
+	 */
+	void handleSlotReportFromTaskManager(final SlotReport slotReport) {
+		runAsync(new Runnable() {
+			@Override
+			public void run() {
+				// TODO slotManager.updateSlot(slotReport);
+			}
+		});
+
+	}
+
+
+	/**
+	 * callback method when current resourceManager is granted leadership
+	 * @param newLeaderSessionID unique leadershipID
+	 */
+	void handleGrantLeadership(final UUID newLeaderSessionID) {
+		runAsync(new Runnable() {
+			@Override
+			public void run() {
+				log.info("ResourceManager {} was granted leadership with leader session ID {}", getAddress(), newLeaderSessionID);
+				leaderSessionID = newLeaderSessionID;
+				// confirming the leader session ID might be blocking, thus do it concurrently
+				getRpcService().scheduleRunnable(
```

Comment:
`getRpcService.runAsync(...)` is equivalent to the `scheduleRunnable` call with `0` delay.